mqtt_client.c

```diff
@@ -745,21 +748,16 @@ esp_err_t esp_mqtt_client_stop(esp_mqtt_client_handle_t client)
     return ESP_OK;
 }
 
-static int esp_mqtt_client_ping(esp_mqtt_client_handle_t client)
+static esp_err_t esp_mqtt_client_ping(esp_mqtt_client_handle_t client)
 {
-    if (client->state != MQTT_STATE_CONNECTED) {
```

Comment:
I feel this should not be removed , because the `MQTT_STATE_ERROR` is not yet used anywhere , but it might be in the future , and this guard would catch it .

Reply:
If the esp_mqtt_task is the only task that can change the client's state, then you know the state the client is in in the moment this function is being called and this code is superfluous (would you try pinging the peer if you are in state MQTT_STATE_ERROR?). If the esp_mqtt_task is not the only task that can change the client's state, then you need a lock around the code that uses the assumption that the client is in some special state, otherwise the moment you run any code using this information can the client be in any different state. This may not seem so critical in a ping sending code, but i haven't seen any locks anywhere else.

Anyway, i don't know anything abut the plans or past of development of this library. Please feel free to introduce me the intentions with MQTT_STATE_ERROR or anything related. This change may indeed go in a wrong direction. Thanks.

Reply:
I am just an user of this component, so I don't know of any plans to extend it.
I actually had my code reach this guard condition when I used an incorrectly initialized client object (I was recreating the client and I missed to make it volatile, and the compiler was optimizing stuff away).
I also see the reasoning behind your argument, that in proper workflow this will not be hit ...

Reply:
Currently the ping function is only called in `MQTT_STATE_CONNECTED` state, so I think the removal is reasonable. Thanks